src/agent/MessageSender.ts

```diff
@@ -33,23 +34,50 @@ export class MessageSender {
     return this._outboundTransporter
   }
 
-  public async packMessage(outboundMessage: OutboundMessage): Promise<OutboundPackage> {
+  public async packMessage(outboundMessage: OutboundMessage, service: DidCommService): Promise<OutboundPackage> {
     const { connection, payload } = outboundMessage
     const { verkey, theirKey } = connection
-    const endpoint = this.transportService.findEndpoint(connection)
+    const endpoint = service.serviceEndpoint
     const message = payload.toJSON()
     this.logger.debug('outboundMessage', { verkey, theirKey, message })
+    const keys = {
+      recipientKeys: service.recipientKeys,
+      routingKeys: service.routingKeys || [],
+      senderVk: connection.verkey,
+    }
+    const wireMessage = await this.envelopeService.packMessage(keys, outboundMessage.payload)
     const responseRequested = outboundMessage.payload.hasReturnRouting()
-    const wireMessage = await this.envelopeService.packMessage(outboundMessage)
     return { connection, payload: wireMessage, endpoint, responseRequested }
   }
 
   public async sendMessage(outboundMessage: OutboundMessage): Promise<void> {
     if (!this.outboundTransporter) {
       throw new AriesFrameworkError('Agent has no outbound transporter!')
     }
-    const outboundPackage = await this.packMessage(outboundMessage)
-    outboundPackage.session = this.transportService.findSession(outboundMessage.connection.id)
-    await this.outboundTransporter.sendMessage(outboundPackage)
+
+    const services = this.transportService.findServices(outboundMessage.connection)
+    if (services.length === 0) {
+      throw new AriesFrameworkError(`Connection with id ${outboundMessage.connection.id} has no service!`)
+    }
+
+    let success = false
+    for await (const service of services) {
+      this.logger.debug(`Sending outbound message to service:`, { service })
+      if (!success) {
+        try {
+          const outboundPackage = await this.packMessage(outboundMessage, service)
+          outboundPackage.session = this.transportService.findSession(outboundMessage.connection.id)
+          await this.outboundTransporter.sendMessage(outboundPackage)
+          success = true
+        } catch (error) {
+          this.logger.debug(
+            `Sending outbound message to service with id ${service.id} failed with the following error:`,
+            {
+              error,
+            }
```

Comment:
This can result in cryptic errors, also adding the message can help
```suggestion
            {
              error,
              errorMessage: error.message
            }
```

Reply:
Ok, no problem. What do you mean by "cryptic error", do you have an example? I would assume that you can read the message from the error object 🤔

Reply:
I think this is the case in Electron and React Native. In electron because we transfer the error over the bridge, meaning it is transformed from an IndyError to a general Error. In React Native (I think) because we're not actually creating an error but rather an object that looks like an error (https://github.com/AbsaOSS/rn-indy-sdk/blob/master/src/index.js#L873-L908, we should fix this)

Reply:
Maybe the things I mentioned are totally unrelated. But I've had trouble with the error message appearing until I explicitly used error.message